compiler/rustc_typeck/src/check/method/suggest.rs

```diff
@@ -1408,81 +1409,159 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
-                        ("an", "Option", "None", ret_ty_matches(sym::Option))
-                    } else {
-                        return;
-                    };
-                if question {
+            // Target wrapper types - types that wrap or pretend to wrap another type,
+            // perhaps this inner type is meant to be called?
+            ty::AdtKind::Struct | ty::AdtKind::Union => {
+                let [first] = ***substs else { return; };
+                let ty::GenericArgKind::Type(ty) = first.unpack() else { return; };
+                let Ok(pick) = self.lookup_probe(
+                            span,
+                            item_name,
+                            ty,
+                            call_expr,
+                            ProbeScope::AllTraits,
+                        )  else { return; };
+
+                let name = self.ty_to_value_string(actual);
+                let inner_id = kind.did();
+
+                if tcx.is_diagnostic_item(sym::LocalKey, inner_id) {
+                    err.help("use `with` or `try_with` to access the contents of threadlocals");
+                } else if Some(kind.did()) == tcx.lang_items().maybe_uninit() {
+                    err.help(format!(
+                        "if this `{name}` has been initialized, \
+                        use one of the `assume_init` methods to access the inner value"
+                    ));
+                } else if tcx.is_diagnostic_item(sym::RefCell, inner_id) {
+                    match pick.autoref_or_ptr_adjustment {
+                        Some(AutorefOrPtrAdjustment::Autoref {
+                            mutbl: Mutability::Not, ..
+                        }) => {
+                            err.span_suggestion_verbose(
+                                expr.span.shrink_to_hi(),
+                                format!(
+                                    "use `.borrow()` to borrow the {ty}, \
+                                panicking if any outstanding mutable borrows exist."
+                                ),
+                                ".borrow()",
+                                Applicability::MaybeIncorrect,
+                            );
+                        }
+                        Some(AutorefOrPtrAdjustment::Autoref {
+                            mutbl: Mutability::Mut, ..
+                        }) => {
+                            err.span_suggestion_verbose(
+                                expr.span.shrink_to_hi(),
+                                format!(
+                                    "use `.borrow_mut()` to mutably borrow the {ty}, \
+                                panicking if any outstanding borrows exist."
+                                ),
+                                ".borrow_mut()",
+                                Applicability::MaybeIncorrect,
+                            );
+                        }
```

Comment:
This could technically be slightly shortened by matching on `mutbl` after the first match. Can you try to see if that makes things a bit shorter?